packages/director/src/app.ts

```diff
@@ -27,18 +30,29 @@ app.get('/', (_, res) =>
 
 app.post('/runs', async (req, res) => {
   const { recordKey, ciBuildId } = req.body;
+  const executionDriver = app.get('executionDriver');
 
   console.log(`>> Machine is asking to join a run`, { recordKey, ciBuildId });
 
-  if(!isKeyAllowed(recordKey)) {
-    console.log(`<< Record key is not allowed`, { recordKey })
+  if (!isKeyAllowed(recordKey)) {
+    console.log(`<< Record key is not allowed`, { recordKey });
 
-    return res.status(403).send(`Provided record key '${recordKey}' is not allowed`);
+    return res
+      .status(403)
+      .send(`Provided record key '${recordKey}' is not allowed`);
   }
 
   console.log(`>> Machine is joining a run`, { ciBuildId });
 
-  const response = await app.get('executionDriver').createRun(req.body);
+  const response = await executionDriver.createRun(req.body);
+  const runWithSpecs = await executionDriver.getRunWithSpecs(response.runId);
```

Comment:
I would say that's a classic use case for a pub/sub

```ts
pub.emit(hookEvents.RUN_START, { runId: response.runId });
```

Reply:
This is probably personal opinion. In my early days of programming I delt with the event emmiter pattern a lot and grew a major distaste for it. Particularly when stuff goes wrong and your emitter is hooked up to multiple subscribers. Tracking down subscriber order is a massive pain. Also just putting break points and trying to debug into a subscriber is kinda succy. Since event emitters push things to the back of the call stack.

If your not using the built in event system its not as big of a deal. Cause you can follow the code flow a little better. But I generally shy away from it cause I find it more understandable to track direct function calls.

That said if you would like me to change it I still can.

Reply:
I agree that it can go left when abused - there are pros and cons. I did see emitters working nicely when controlled and used strictly. I also suffered a lot while working on systems that abused the pattern and lost control of emitters / listeners. 

Bottom line, if you dislike emitters so much, let's try to find a different solution. What I am looking for is to separate the functionality that deals with director primary responsibility (manage the tests) from hooks functionality.

Simply extracting all hooks functionality from `app.ts` into a separate module would also work for me.